docs(build): Add ECC dependencies documentation (#7794)
ZcashFoundation/zebra: book/src/dev/ecc-updates.md

## Steps for upgrading

Let's dive into the details of each step required to perform an upgrade:

### Before starting

- Zebra developers often dismiss ECC dependency upgrade suggestions from dependabot. For instance, see [this closed PR](https://github.com/ZcashFoundation/zebra/pull/7745) in favor of the [5.7.0 zcashd upgrade PR](https://github.com/ZcashFoundation/zebra/pull/7784), which followed this guide.

- Determine the version of `zcashd` to use. This version will determine which versions of other crates to use. Typically, this should be a [tag](https://github.com/zcash/zcash/tags), but in some cases, it might be a reference to a branch (e.g., nu5-consensus) for testing unreleased developments.

- Upgrading the `zcash_script` crate can be challenging, depending on changes in the latest `zcashd` release. Follow the instructions in the project's [README](https://github.com/ZcashFoundation/zcash_script/blob/master/README.md) for guidance.

- If you haven't upgraded `zcash_script`, don't proceed with upgrading other ECC dependencies in Zebra.

> - Ensure that the crate versions in the `Cargo.toml` of the zcashd release, `Cargo.toml` of `zcash_script`, and the `Cargo.toml` files of Zebra crates are all the same. Version consistency is crucial.

Reviewer: Optional: Suggest checking cargo deny with a locally upgraded version of zcash_script before publishing like the zcash_script readme?

Author: Yea, good idea. I added it to the deny.toml section more at the bottom of the file: https://github.com/ZcashFoundation/zebra/pull/7794/commits/990d6082645ce869e5b7537d7480a31dcad65c00

Reviewer 2: I wonder if it would also help to run `cargo deny` in the `zcash_script` CI?
Sometimes we forget manual steps.

Author: https://github.com/ZcashFoundation/zcash_script/issues/114

### Upgrade versions

- You can upgrade versions manually or use a search and replace script in all `Cargo.toml` files within Zebra. Alternatively, you can use the `cargo upgrade` command. For example, in [this PR](https://github.com/ZcashFoundation/zebra/pull/7784), the following command was used:

```
cargo upgrade --incompatible -p bridgetree -p incrementalmerkletree -p orchard -p zcash_primitives@0.13.0-rc.1 -p zcash_proofs@0.13.0-rc.1 -p zcash_script
```

Notes:

- Insert all the crate names to be updated to the command.

- We can upgrade to specific versions in some crates instead of just the last one (no version specified). 

- You need to have [cargo upgrade](https://crates.io/crates/cargo-upgrades) and [cargo edit](https://crates.io/crates/cargo-edit) installed for this command to work.

### Build/Test zebra & fix issues

- Build zebra and make sure it compiles.

```
cargo build
```

- Test Zebra and make sure all test code compiles and all tests pass:

```
cargo test
```

> - When upgrading, it's common for things to break, such as deprecated functions or removed functionality. Address these issues by referring to the crate's changelog, which often provides explanations and workarounds.

Reviewer: Optional/Nitpick:

```suggestion
- When upgrading, it's common for things to break, such as deprecated or removed functionality. Address these issues by referring to the crate's changelog, which often provides explanations and workarounds.
```

Author: https://github.com/ZcashFoundation/zebra/pull/7794/commits/4cf4136b98d0e36a7b2e211658642e1cefe6abe9

## Steps for upgrading

Let's dive into the details of each step required to perform an upgrade:

### Before starting

- Zebra developers often dismiss ECC dependency upgrade suggestions from dependabot. For instance, see [this closed PR](https://github.com/ZcashFoundation/zebra/pull/7745) in favor of the [5.7.0 zcashd upgrade PR](https://github.com/ZcashFoundation/zebra/pull/7784), which followed this guide.

- Determine the version of `zcashd` to use. This version will determine which versions of other crates to use. Typically, this should be a [tag](https://github.com/zcash/zcash/tags), but in some cases, it might be a reference to a branch (e.g., nu5-consensus) for testing unreleased developments.

- Upgrading the `zcash_script` crate can be challenging, depending on changes in the latest `zcashd` release. Follow the instructions in the project's [README](https://github.com/ZcashFoundation/zcash_script/blob/master/README.md) for guidance.

> - If you haven't upgraded `zcash_script`, don't proceed with upgrading other ECC dependencies in Zebra.

Reviewer 2: Double negatives ("if you haven't, … don't …") are confusing, and it's easy to miss one of the negations.

Instead, say what people should do:
```suggestion
- Upgrade and release `zcash_script` before upgrading other ECC dependencies in Zebra.
```

> - Ensure that the crate versions in the `Cargo.toml` of the zcashd release, `Cargo.toml` of `zcash_script`, and the `Cargo.toml` files of Zebra crates are all the same. Version consistency is crucial.

Reviewer 2: When is this version consistency check done?
If it is done before starting, then it will always be true, because nothing has been upgraded yet.

Does it belong at the top of the "Upgrade versions" section instead?

Author: https://github.com/ZcashFoundation/zebra/pull/7794/commits/84ba7c728c47d97523af9c425e1c3ca850079792

### Upgrade versions

> - You can upgrade versions manually or use a search and replace script in all `Cargo.toml` files within Zebra. Alternatively, you can use the `cargo upgrade` command. For example, in [this PR](https://github.com/ZcashFoundation/zebra/pull/7784), the following command was used:

Reviewer 2: Search and replace won't work reliably if some of the dependencies use features or are optional, because they use an expanded format. Similarly, using a manual replace is unreliable.

If version consistency is crucial, using a reliable tool is the best way.

```suggestion
- Use the `cargo upgrade` command to upgrade all the ECC dependency versions in Zebra. For example, in [this PR](https://github.com/ZcashFoundation/zebra/pull/7784), the following command was used:
```

```

> cargo upgrade --incompatible -p bridgetree -p incrementalmerkletree -p orchard -p zcash_primitives@0.13.0-rc.1 -p zcash_proofs@0.13.0-rc.1 -p zcash_script

Reviewer 2: Let's use an example that will work for the next upgrade as well?

There are more crates than this, but these are the Zcash-specific ones, the others are cryptographic and we sometimes upgrade them independently.

```suggestion
cargo upgrade --incompatible -p bridgetree -p incrementalmerkletree -p orchard -p zcash_primitives -p zcash_proofs -p zcash_address -p zcash_encoding -p zcash_note_encryption -p zcash_script
```

```

Notes:

- Insert all the crate names to be updated to the command.

> - We can upgrade to specific versions in some crates instead of just the last one (no version specified). 

Reviewer 2: Since I deleted the specific version upgrades, I said how to do them here. The default is the highest version. (The last version released can sometimes be a patch release on an earlier major version.)

```suggestion
- Use `crate-name@version` to upgrade to a specific version of that crate, instead of just the highest version. 
```

- You need to have [cargo upgrade](https://crates.io/crates/cargo-upgrades) and [cargo edit](https://crates.io/crates/cargo-edit) installed for this command to work.

### Build/Test zebra & fix issues

- Build zebra and make sure it compiles.

```
cargo build
```

- Test Zebra and make sure all test code compiles and all tests pass:

```
cargo test
```

> - When upgrading, it's common for things to break, such as deprecated or removed functionality. Address these issues by referring to the crate's changelog, which often provides explanations and workarounds.

Reviewer 2: Which crate's changelog?

```suggestion
- When upgrading, it's common for things to break, such as deprecated or removed functionality. Address these issues by referring to the broken dependency's changelog, which often provides explanations and workarounds.
```
